src/openfermion/ops/_code_operator.py

```diff
@@ -0,0 +1,298 @@
+    input of length N, which has entries either of type SymbolicBinary, or of 
+    valid type for an input of the SymbolicBinary-constructor. 
+    
+    The signs + and *, += and *= are overloaded to implement concatenation
+    and appendage on BinaryCode-objects.
+
+    NOTE: that multiplication of a BinaryCode with an integer yields a
+    multiple appending of the same code, the multiplication with another
+    BinaryCode their concatenation.
+    """
+
+    def __init__(self, encoding, decoding):
+        """ Initialization of a binary code.
+
+        Args: 
+            encoding (array or list): nested lists or binary 2D-array
+            decoding (array or list): list of SymbolicBinary, list-like or str
+
+        Raises: 
+            TypeError: non-list, array like encoding or decoding, unsuitable
+            symbolicBinary generators,
+            BinaryCodeError: in case of decoder/encoder size mismatch or decoder
+            size, qubits indexed mismatch
+        """
+
+        if not isinstance(encoding, (numpy.ndarray, list, tuple)):
+            raise TypeError('encoding must be a list, array or tuple .')
+
+        if not isinstance(decoding, (numpy.ndarray, list, tuple)):
+            raise TypeError('decoding must be a list, array or tuple .')
+
+        self.enc = scipy.sparse.csc_matrix(encoding)
+        self.qubits, self.orbitals = numpy.shape(encoding)
+
+        if (self.orbitals != len(decoding)):
+            raise BinaryCodeError(
+                'size mismatch, decoder and encoder should have the same'
+                ' first dimension')
+
+        decoder_qubits = set()
+        self.dec = []
+
+        for symbolic_binary in numpy.array(decoding):
+
+            if isinstance(symbolic_binary, (tuple, list, str)):
+                symbolic_binary = SymbolicBinary(symbolic_binary)
+            if isinstance(symbolic_binary, SymbolicBinary):
+                self.dec.append(symbolic_binary)
+                decoder_qubits = decoder_qubits | set(
+                    symbolic_binary.enumerate_qubits())
+            else:
+                raise TypeError(
+                    'decoder component provided '
+                    'is not a suitable for SymbolicBinary',
+                    symbolic_binary)
+
+        if len(decoder_qubits) != self.qubits:
+            raise BinaryCodeError(
+                'decoder and encoder provided has different number of qubits')
+
+        if max(decoder_qubits) + 1 > self.qubits:
+            raise BinaryCodeError('decoder is not indexing some qubits. Qubits'
+                                  'indexed are: {}'.format(decoder_qubits))
+
+        self.dec = numpy.array(self.dec)
+
+    def __iadd__(self, appendix):
+        """ In-place appending a binary code with +=.
+        
+        Args:
+            appendix (BinaryCode): The code to append to the present one. 
+
+        Returns (BinaryCode): a global binary code with size n1+n2, N1+N2
```

Comment:
I think if you are going to refer to size of encoded and unencoded spaces as "n" and "N" as opposed to something like "encoded_dimension" and "unencoded_dimension" then I think you need to keep reminding people what "n" and "N" are.